From d9f943b9747d39416a5febaf00fc20747abfd618 Mon Sep 17 00:00:00 2001
From: Roland Haas <roland.haas@physics.gatech.edu>
Date: Sat, 14 Jul 2012 16:09:25 -0700
Subject: [PATCH 3/3] GRHydro: add routines to access
 CarpetEvolutionMask::evolution_mask

based on runtime parameters, ie. CCTK_VarDataPtr for Fortran
---
 param.ccl                     |    4 ++
 src/GRHydro_Con2Prim.F90      |   13 ++++
 src/GRHydro_EvolutionMask.F90 |  143 +++++++++++++++++++++++++++++++++++++++++
 src/GRHydro_EvolutionMask.c   |   83 ++++++++++++++++++++++++
 src/make.code.defn            |    4 +-
 5 files changed, 246 insertions(+), 1 deletion(-)
 create mode 100644 src/GRHydro_EvolutionMask.F90
 create mode 100644 src/GRHydro_EvolutionMask.c

diff --git a/param.ccl b/param.ccl
index a89d8cc..4cfeb1d 100644
--- a/param.ccl
+++ b/param.ccl
@@ -66,6 +66,10 @@ USES INT spatial_order
 
 restricted:
 
+BOOLEAN use_evolution_mask "Set this to 'true' to skip validity tests in regions where CarpetEvolutionMask::evolution_mask vanishes." STEERABLE=ALWAYS
+{
+} "false"
+
 BOOLEAN GRHydro_enable_internal_excision "Set this to 'false' to disable the thorn-internal excision." STEERABLE=ALWAYS
 {
 } "true"
diff --git a/src/GRHydro_Con2Prim.F90 b/src/GRHydro_Con2Prim.F90
index 9f5e511..920b98d 100644
--- a/src/GRHydro_Con2Prim.F90
+++ b/src/GRHydro_Con2Prim.F90
@@ -2210,6 +2210,7 @@ end subroutine sync_GRHydro_C2P_failed
 subroutine check_GRHydro_C2P_failed(CCTK_ARGUMENTS)
   
   use Con2Prim_fortran_interfaces
+  use GRHydro_EvolutionMask
 
   implicit none
   
@@ -2223,6 +2224,9 @@ subroutine check_GRHydro_C2P_failed(CCTK_ARGUMENTS)
   integer :: i, j, k, nx, ny, nz
   character(len=100) warnline
 
+  ! we skip points where evolution_mask vanishes
+  CCTK_REAL, DIMENSION(:,:,:), POINTER :: evolution_mask
+
   ! save memory when MP is not used
   CCTK_INT :: GRHydro_UseGeneralCoordinates
   CCTK_REAL, DIMENSION(:,:,:,:), POINTER :: vup
@@ -2233,6 +2237,8 @@ subroutine check_GRHydro_C2P_failed(CCTK_ARGUMENTS)
     vup => vel
   end if
 
+  call GRHydro_DeclareEvolutionMask(cctkGH, evolution_mask)
+
 !  call CCTK_INFO("Checking the C2P failure mask.")
 
   nx = cctk_lsh(1)
@@ -2260,6 +2266,13 @@ subroutine check_GRHydro_C2P_failed(CCTK_ARGUMENTS)
                cycle
            end if 
 
+          ! do not collapse conditions since Fortran does not guarantee an order
+          if (use_evolution_mask.ne.0) then
+            if (evolution_mask(i,j,k).eq.0d0) then 
+              cycle
+            end if
+          end if
+
           !$OMP CRITICAL
           call CCTK_WARN(1,'Con2Prim failed; stopping the code.')
           call CCTK_WARN(1,'Even with mesh refinement, this point is not restricted from a finer level, so this is really an error')
diff --git a/src/GRHydro_EvolutionMask.F90 b/src/GRHydro_EvolutionMask.F90
new file mode 100644
index 0000000..9acb18e
--- /dev/null
+++ b/src/GRHydro_EvolutionMask.F90
@@ -0,0 +1,143 @@
+/*@@
+   @file      GRHydro_EvolutionMask.F90
+   @date      Sat Jul 14 15:38:02 PDT 2012
+   @author    Roland Haas
+   @desc 
+   User level module and Fortran glue code to get access to
+   CarpetEvolutionMask::evolution_mask based on runtime parameters.
+   @enddesc 
+ @@*/
+#include "cctk.h"
+#include "cctk_Parameters.h"
+
+ /*@@
+   @routine    GRHydro_EvolutionMaskInternal
+   @date       Sat Jul 14 15:39:37 PDT 2012
+   @author     Roland Haas
+   @desc
+   Internal module used only to transport a pointer from C to Fortran
+   @history
+
+   @endhistory
+
+@@*/
+module GRHydro_EvolutionMaskInternal
+
+      implicit none
+
+      CCTK_REAL, pointer, dimension(:,:,:) :: evolution_mask_i
+end module
+
+ /*@@
+   @routine    GRHydro_EvolutionMaskCopyPointer
+   @date       Sat Jul 14 15:39:37 PDT 2012
+   @author     Roland Haas
+   @desc
+   Stores a pointer to a Fortran array in the Fotran pointer evolution_mask_i
+   @enddesc
+   @calls     
+   @calledby
+   @history
+
+   @endhistory
+   @var     evolution_mask_
+   @vdesc   Fortran array
+   @vtype   CCTK_REAL, dimension(cctk_lsh(1), cctk_lsh(2), cctk_lsh(3))
+   @vio     in
+   @vcomment
+   @endvar
+
+   @var     cctk_ash
+   @vdesc   allocated array size
+   @vtype   integer, dimension(3)
+   @vio     in
+   @vcomment
+   @endvar
+
+   @returntype none
+   @returndesc
+   @endreturndesc
+@@*/
+subroutine GRHydro_EvolutionMaskCopyPointer(evolution_mask_, cctk_ash)
+
+  use GRHydro_EvolutionMaskInternal
+
+  implicit none
+
+  integer, dimension(3) :: cctk_ash
+  CCTK_REAL, dimension(cctk_ash(1), cctk_ash(2), cctk_ash(3)), target :: evolution_mask_
+
+  evolution_mask_i => evolution_mask_
+
+end subroutine
+
+
+ /*@@
+   @routine    GRHydro_EvolutionMask
+   @date       Sat Jul 14 15:39:37 PDT 2012
+   @author     Roland Haas
+   @desc
+   User level module to get access to CarpetEvolutionMask::evolution_mask based
+   on runtime parameters.
+   @history
+
+   @endhistory
+
+@@*/
+module GRHydro_EvolutionMask
+
+      implicit none
+
+CONTAINS
+       /*@@
+         @routine    GRHydro_DeclareEvolutionMask
+         @date       Sat Jul 14 15:39:37 PDT 2012
+         @author     Roland Haas
+         @desc
+         Stores a pointer to CarpetEvolutionMask::evoltuion_mask in its
+         argument.
+         @enddesc
+         @calls     
+         @calledby
+         @history
+      
+         @endhistory
+
+         @var     cctkGH
+         @vdesc   Cactus grid hierarchy
+         @vtype   cGH *
+         @vio     in
+         @vcomment
+         @endvar
+
+         @var     evolution_mask
+         @vdesc   Fortran array pointer
+         @vtype   CCTK_REAL, dimension(:,:,:), pointer
+         @vio     in
+         @vcomment
+         @endvar
+      
+         @returntype none
+         @returndesc
+         @endreturndesc
+      @@*/
+      subroutine GRHydro_DeclareEvolutionMask(cctkGH, evolution_mask)
+
+        use GRHydro_EvolutionMaskInternal
+
+        implicit none
+
+        DECLARE_CCTK_PARAMETERS
+
+        CCTK_POINTER_TO_CONST :: cctkGH
+        CCTK_REAL, pointer, dimension(:,:,:) :: evolution_mask
+
+        if(use_evolution_mask .ne. 0) then
+          call GRHydro_DeclareEvolutionMaskC(cctkGH)
+          evolution_mask => evolution_mask_i
+        else
+          Nullify(evolution_mask)
+        end if
+
+      end subroutine
+end module
diff --git a/src/GRHydro_EvolutionMask.c b/src/GRHydro_EvolutionMask.c
new file mode 100644
index 0000000..65c718b
--- /dev/null
+++ b/src/GRHydro_EvolutionMask.c
@@ -0,0 +1,83 @@
+/*@@
+   @file      GRHydro_EvolutionMask.c
+   @date      Sat Jul 14 15:38:02 PDT 2012
+   @author    Roland Haas
+   @desc 
+   Internal routine to get access to CarpetEvolutionMask::evolution_mask based
+   on runtime parameters.
+   @enddesc 
+ @@*/
+
+#include "cctk.h"
+#include "cctk_Arguments.h"
+#include "cctk_Parameters.h"
+
+/***********************************************************************/
+/*********** External function prototypes ******************************/
+/***********************************************************************/
+extern void CCTK_FNAME(GRHydro_EvolutionMaskCopyPointer)(
+    CCTK_POINTER_TO_CONST vardataptr, const int cctk_ash[3]);
+
+/***********************************************************************/
+/*********** Externally visible routines *******************************/
+/***********************************************************************/
+void CCTK_FCALL CCTK_FNAME(GRHydro_DeclareEvolutionMaskC)(
+    const cGH ** p_cctkGH);
+
+ /*@@
+   @routine    GRHydro_DeclareEvolutionMaskC
+   @date       Sat Jul 14 15:39:37 PDT 2012
+   @author     Roland Haas
+   @desc
+   Provides the equivalent of CCTK_VarDataPtr to Fortran.
+   @enddesc
+   @calls     GRHydro_EvolutionMaskCopyPointer
+   @calledby
+   @history
+
+   @endhistory
+   @var     p_cctkGH
+   @vdesc   grid hierarchy pointer
+   @vtype   cGH **
+   @vio     in
+   @vcomment
+
+   @endvar
+
+   @returntype none
+   @returndesc
+   @endreturndesc
+@@*/
+void CCTK_FCALL CCTK_FNAME(GRHydro_DeclareEvolutionMaskC)(const cGH **p_cctkGH)
+{
+  DECLARE_CCTK_PARAMETERS;
+
+  const cGH *cctkGH = *p_cctkGH;
+  static int varindex = -1;
+  const CCTK_REAL *vardataptr;
+
+  // use Fortran glue routines to insert pointer to
+  // CarpetEvolutionMask::evolution_mask into the module GRHydro_EvolutionMask
+  if(varindex < 0)
+  {
+    varindex = CCTK_VarIndex("CarpetEvolutionMask::evolution_mask");
+    if(varindex < 0)
+    {
+      CCTK_VWarn(CCTK_WARN_ABORT, __LINE__, __FILE__, CCTK_THORNSTRING,
+          "Could not obtain variable index for '%s': %d",
+          "CarpetEvolutionMask::evolution_mask", varindex);
+      return; // NOTREACHED
+    }
+  }
+
+  vardataptr = CCTK_VarDataPtrI(cctkGH, 0, varindex);
+  if(vardataptr == NULL)
+  {
+    CCTK_VWarn(CCTK_WARN_ABORT, __LINE__, __FILE__, CCTK_THORNSTRING,
+        "Could not obtain data pointer for '%s'.",
+        "CarpetEvolutionMask::evolution_mask");
+    return; // NOTREACHED
+  }
+
+  CCTK_FNAME(GRHydro_EvolutionMaskCopyPointer)(vardataptr, cctkGH->cctk_ash);
+}
diff --git a/src/make.code.defn b/src/make.code.defn
index 25b0aa1..b448662 100644
--- a/src/make.code.defn
+++ b/src/make.code.defn
@@ -63,7 +63,9 @@ SRCS = 	Utils.F90 \
 	GRHydro_ENOReconstruct_drv.F90\
 	GRHydro_PPMReconstruct_drv.F90\
 	GRHydro_LastMoLPostStep.c\
-	GRHydro_TVDReconstruct_drv.F90
+	GRHydro_TVDReconstruct_drv.F90\
+	GRHydro_EvolutionMask.F90\
+	GRHydro_EvolutionMask.c\
 
 
 
-- 
1.7.10.4

